File: Leetcode/125.go
package Leetcode

import (
	"unicode"
)

func isPalindrome(s string) bool {
	n := len(s)

	for head, tail := 0, n-1; head < tail; {
		for head < tail && !isAlphaNumeric(rune(s[head])) {
			head++
		}
		for head < tail && !isAlphaNumeric(rune(s[tail])) {
			tail--
		}
		if head == tail {
			return true
		}
		if unicode.ToLower(rune(s[head])) != unicode.ToLower(rune(s[tail])) {
			return false
		}
		head++
		tail--
	}
	return true
}

func isAlphaNumeric(ch rune) bool {
	return unicode.IsDigit(ch) || unicode.IsLetter(ch)
}

File: Leetcode/125.py
class Solution:
    def isPalindrome(self, s: str) -> bool:
        n = len(s)
        head, tail = 0, n - 1
        while head < tail:
            while head < tail and not s[head].isalnum():
                head += 1
            while head < tail and not s[tail].isalnum():
                tail -= 1
            if head == tail:
                return True
            if s[head].lower() != s[tail].lower():
                return False
            head += 1
            tail -= 1
        return True
